### PR TITLE
Add DomainEventBus and use it to automatically fetch when a repo is connected

### PR DESCRIPTION
**Pull request by mattwynne**
Inspired by https://rjzaworski.com/2019/10/event-emitters-in-typescript, implements a simple `DomainEventBus` and a couple of `DomainEvent`s, then uses them to automatically run a fetch when a repo's origin is set.

I've invented a `Rule` abstraction that could model other feedback loops that react to domain events and run commands through the application layer. I guess this is a bit like what we called a "Saga" in Jam?

Even though they're not used in there, I made the rules a constructor parameter to the app, so that anyone using the app has to think about what rules they want to use. Otherwise it would be too easy to forget them in `server.ts`

Questions:
- Does it make sense?
- Are we happy with the `Rule` name?
- Will the `Rule`s stay in memory OK given they're not held on a field in `LaBoîte`?
- Any other reactions?

I know this won't cope very well if the fetch fails, but that's a bigger problem. We might need to start adding persistence somewhere so we can store the results of these kinds of processes that happen outside of a users's request.

### CODE REVIEW

**Comment by mattwynne** [comment]
Another thought, for the purist (I'm looking at you @jbpros):

Should the `Repo` be given the whole `DomainEventBus` interface, or a subset that just includes the `emit` function? i.e. should domain objects be able to subscribe to domain events themselves? I am inclined to think not, but I can't really explain why.

**Comment by mattwynne** [comment]
On reflection, I thnk the `remoteUrl` should not leak out in a domain event. Maybe it's better to call that event `repo.connected`

File: packages/app/features/fetch_after_connection.feature [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Feature: Fetch after connection
+
+  After a repo is connected, we automatically run a fetch in the
+  background.
+
+  Scenario: Succesful fetch after a new connection
+    Given a remote repo with commits on the "main" branch
+    When a consumer connects the remote repo
+    Then the repo should be fetched
```
Comment (jbpros) [inline packages/app/features/fetch_after_connection.feature:9]: looks like your IDE is not properly configured (?)

Comment (mattwynne) [reply packages/app/features/fetch_after_connection.feature:9]: In what way? Whitespace?

Comment (jbpros) [reply packages/app/features/fetch_after_connection.feature:9]: It's not adding a newline character at the end of the last line.

Comment (jbpros) [reply packages/app/features/fetch_after_connection.feature:9]: Seems to be in feature files only

**Review by jbpros — COMMENTED** [review]
> Another thought, for the purist (I'm looking at you @jbpros):
> 
> Should the `Repo` be given the whole `DomainEventBus` interface, or a subset that just includes the `emit` function? i.e. should domain objects be able to subscribe to domain events themselves? I am inclined to think not, but I can't really explain why.

Only the `emit` function would be better, imo. We just want the entity to broadcast events when it has done something meaningful to the rest of the world, nothing more. I think that "sagas" would be a better tool to capture long-running/complex behaviour that spans multiple events.

**Comment by jbpros** [comment]
> Even though they're not used in there, I made the rules a constructor parameter to the app, so that anyone using the app has to think about what rules they want to use. Otherwise it would be too easy to forget them in `server.ts`

Not sure I understand the rationale here. I like the idea and at the same time I don't really see how useful that is ;) Is the point to allow for configurable high-level business behaviour? Wouldn't you want most of the "rules" to be always active though?

I think the name `Rule` is not clear enough. Suggestions: `DomainRule`, `BusinessRule`, `DomainProcess`, `BusinessProcess` :)

**Comment by mattwynne** [comment]
> Not sure I understand the rationale here.

Ideally we need a single place, shared between `server.ts` and the acceptance tests, to configure the set of rules to use. I guess if we had some kind of factory/assemble function for the whole application and its dependencies, it could go in there.

> Wouldn't you want most of the "rules" to be always active though?

Yes, I'm just not sure where would be the right place to hard-code the list of them.

> I think the name Rule is not clear enough. Suggestions: DomainRule, BusinessRule, DomainProcess, BusinessProcess

What about `Saga`? Are these different to Sagas? If so, how?

**Pull request closed without merging**